File: jvm/periph-java/src/main/java/it/uhde/periph/chips/motor/DRV8830Full.java
package it.uhde.periph.chips.motor;

import it.uhde.periph.connection.Connection;
import it.uhde.periph.connection.EdgeHandler;
import it.uhde.periph.connection.EdgeTrigger;
import it.uhde.periph.connection.InputPin;

import java.io.IOException;
import java.util.function.Consumer;

/**
 * DRV8830 full interface — extends {@link DRV8830Minimal} with raw
 * {@code CONTROL} access, output read-back, fault reporting/clearing, and the
 * {@code FAULTn} interrupt API.
 *
 * <p>Faults are never cleared implicitly: a latched OCP/ILIMIT fault also
 * disables the H-bridge, so clearing is always an explicit
 * {@link #clearFault()}. {@link #onInterrupt(Consumer)} uses
 * {@code connection.intPin()} if wired, otherwise a 5&nbsp;ms polling thread
 * that fires once per new fault.
 */
public class DRV8830Full extends DRV8830Minimal {

    /** H-bridge state decoded from {@code IN1}/{@code IN2}. */
    public enum Direction {
        /** IN1=0, IN2=0 — outputs high-Z (standby). */
        COAST,
        /** IN1=1, IN2=0. */
        FORWARD,
        /** IN1=0, IN2=1. */
        REVERSE,
        /** IN1=1, IN2=1 — both outputs high. */
        BRAKE
    }

    /**
     * Decoded {@code CONTROL} register.
     *
     * @param voltage   commanded magnitude in V (0 for a reserved {@code VSET} code)
     * @param direction H-bridge state
     */
    public record Output(double voltage, Direction direction) {}

    /**
     * Decoded {@code FAULT} register.
     *
     * @param fault  any fault condition exists
     * @param ocp    overcurrent (short-circuit) event
     * @param uvlo   undervoltage lockout
     * @param ots    overtemperature shutdown
     * @param ilimit extended current-limit event
     */
    public record Fault(boolean fault, boolean ocp, boolean uvlo, boolean ots, boolean ilimit) {}

    private volatile Consumer<Fault> callback;
    private InputPin intPin;
    private volatile boolean polling = false;
    private Thread pollThread;
    private final EdgeHandler edgeHandler = this::handleEdge;

    /**
     * Construct the driver; same presence check as {@link DRV8830Minimal}.
     *
     * @param connection configured I²C connection bound to the device (0x60–0x68)
     * @throws IOException on bus error
     */
    public DRV8830Full(Connection connection) throws IOException {
        super(connection);
    }

    /**
     * Write the {@code CONTROL} register from raw fields.
     *
     * @param vset {@code VSET} DAC code, 6–63 (0–5 are reserved)
     * @param in1  H-bridge input 1
     * @param in2  H-bridge input 2
     * @throws IllegalArgumentException if {@code vset} is outside 6–63
     * @throws IOException on bus error
     */
    public void setOutput(int vset, boolean in1, boolean in2) throws IOException {
        if (vset < VSET_MIN || vset > VSET_MAX) {
            throw new IllegalArgumentException("vset must be 6-63, got " + vset);
        }
        writeReg(REG_CONTROL, (vset << 2) | (in1 ? CTRL_IN1 : 0) | (in2 ? CTRL_IN2 : 0));
    }

    /**
     * Read back and decode the {@code CONTROL} register.
     *
     * @return commanded voltage magnitude and H-bridge direction
     * @throws IOException on bus error
     */
    public Output readOutput() throws IOException {
        int ctrl = readReg(REG_CONTROL);
        return new Output(vsetToVoltage(ctrl >> 2), Direction.values()[ctrl & 0x03]);
    }

    /**
     * Read the {@code FAULT} register without clearing it.
     *
     * @return decoded fault flags
     * @throws IOException on bus error
     */
    public Fault readFault() throws IOException {
        int f = readReg(REG_FAULT);
        return new Fault((f & FAULT_FAULT) != 0, (f & FAULT_OCP) != 0, (f & FAULT_UVLO) != 0,
                (f & FAULT_OTS) != 0, (f & FAULT_ILIMIT) != 0);
    }

    /**
     * Clear all fault status bits ({@code CLEAR} = 1); re-enables the H-bridge
     * if an OCP/ILIMIT fault had latched it off.
     *
     * @throws IOException on bus error
     */
    public void clearFault() throws IOException {
        writeReg(REG_FAULT, FAULT_CLEAR);
    }

    // -------------------------------------------------------------------------
    // Interrupt API (Level 1 — FAULTn)
    // -------------------------------------------------------------------------

    /**
     * Subscribe to fault interrupts using {@code connection.intPin()} (or a
     * 5&nbsp;ms polling thread if none is wired). The fault is not cleared.
     *
     * @param callback called with the {@link #readFault()} result on each new fault
     */
    public void onInterrupt(Consumer<Fault> callback) {
        onInterrupt(callback, connection.intPin());
    }

    /**
     * Subscribe to fault interrupts, overriding which {@link InputPin} delivers edges.
     *
     * @param callback called with the {@link #readFault()} result on each new fault
     * @param intPin   FAULTn pin to arm (falling edge), or {@code null} to force the 5&nbsp;ms polling fallback
     */
    public void onInterrupt(Consumer<Fault> callback, InputPin intPin) {
        offInterrupt();
        this.callback = callback;
        if (intPin != null) {
            this.intPin = intPin;
            intPin.onEdge(edgeHandler, EdgeTrigger.FALLING);
        } else {
            startPolling();
        }
    }

    /** Unsubscribe and stop delivery. */
    public void offInterrupt() {
        callback = null;
        if (intPin != null) {
            intPin.offEdge(edgeHandler);
            intPin = null;
        }
        stopPolling();
    }

    /**
     * Read the fault status — equivalent to {@link #readFault()}, does not clear.
     *
     * @return decoded fault flags
     * @throws IOException on bus error
     */
    public Fault pollInterrupt() throws IOException {
        return readFault();
    }

    private void startPolling() {
        polling = true;
        pollThread = new Thread(() -> {
            boolean wasFault = false;
            while (polling) {
                try {
                    Fault status = pollInterrupt();
                    Consumer<Fault> cb = callback;
                    if (status.fault() && !wasFault && cb != null) cb.accept(status);
                    wasFault = status.fault();
                } catch (IOException ignored) {
                    // bus error; retry on the next tick
                }
                try {
                    Thread.sleep(5);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return;
                }
            }
        }, "drv8830-poll");
        pollThread.setDaemon(true);
        pollThread.start();
    }

    private void stopPolling() {
        polling = false;
        if (pollThread != null) {
            pollThread.interrupt();
            pollThread = null;
        }
    }

    private void handleEdge() {
        try {
            Fault status = pollInterrupt();
            Consumer<Fault> cb = callback;
            if (status.fault() && cb != null) cb.accept(status);
        } catch (IOException ignored) {
            // bus error; wait for the next edge rather than propagating
        }
    }
}

File: jvm/periph-kotlin/src/main/kotlin/it/uhde/periph/chips/motor/DRV8830Full.kt
package it.uhde.periph.chips.motor

import it.uhde.periph.connection.Connection
import it.uhde.periph.connection.EdgeHandler
import it.uhde.periph.connection.EdgeTrigger
import it.uhde.periph.connection.InputPin
import java.io.IOException

/**
 * DRV8830 full interface — extends [DRV8830Minimal] with raw `CONTROL` access,
 * output read-back, fault reporting/clearing, and the `FAULTn` interrupt API.
 *
 * Faults are never cleared implicitly: a latched OCP/ILIMIT fault also
 * disables the H-bridge, so clearing is always an explicit [clearFault].
 * [onInterrupt] uses `connection.intPin()` if wired, otherwise a 5 ms polling
 * thread that fires once per new fault.
 *
 * @param connection configured I²C connection bound to the device (0x60–0x68)
 */
class DRV8830Full(connection: Connection) : DRV8830Minimal(connection) {

    /** H-bridge state decoded from `IN1`/`IN2`. */
    enum class Direction {
        /** IN1=0, IN2=0 — outputs high-Z (standby). */
        COAST,
        /** IN1=1, IN2=0. */
        FORWARD,
        /** IN1=0, IN2=1. */
        REVERSE,
        /** IN1=1, IN2=1 — both outputs high. */
        BRAKE,
    }

    /**
     * Decoded `CONTROL` register.
     *
     * @property voltage commanded magnitude in V (0 for a reserved `VSET` code)
     * @property direction H-bridge state
     */
    data class Output(val voltage: Double, val direction: Direction)

    /**
     * Decoded `FAULT` register.
     *
     * @property fault any fault condition exists
     * @property ocp overcurrent (short-circuit) event
     * @property uvlo undervoltage lockout
     * @property ots overtemperature shutdown
     * @property ilimit extended current-limit event
     */
    data class Fault(val fault: Boolean, val ocp: Boolean, val uvlo: Boolean, val ots: Boolean, val ilimit: Boolean)

    @Volatile private var callback: ((Fault) -> Unit)? = null
    private var intPin: InputPin? = null
    @Volatile private var polling: Boolean = false
    private var pollThread: Thread? = null
    private val edgeHandler = EdgeHandler { handleEdge() }

    /**
     * Write the `CONTROL` register from raw fields.
     *
     * @param vset `VSET` DAC code, 6–63 (0–5 are reserved)
     * @param in1 H-bridge input 1
     * @param in2 H-bridge input 2
     * @throws IllegalArgumentException if [vset] is outside 6–63
     */
    fun setOutput(vset: Int, in1: Boolean, in2: Boolean) {
        require(vset in VSET_MIN..VSET_MAX) { "vset must be 6-63, got $vset" }
        writeReg(REG_CONTROL, (vset shl 2) or (if (in1) CTRL_IN1 else 0) or (if (in2) CTRL_IN2 else 0))
    }

    /** Read back and decode the `CONTROL` register. */
    fun readOutput(): Output {
        val ctrl = readReg(REG_CONTROL)
        return Output(vsetToVoltage(ctrl shr 2), Direction.entries[ctrl and 0x03])
    }

    /** Read the `FAULT` register without clearing it. */
    fun readFault(): Fault {
        val f = readReg(REG_FAULT)
        return Fault(
            (f and FAULT_FAULT) != 0, (f and FAULT_OCP) != 0, (f and FAULT_UVLO) != 0,
            (f and FAULT_OTS) != 0, (f and FAULT_ILIMIT) != 0,
        )
    }

    /** Clear all fault status bits (`CLEAR` = 1); re-enables the H-bridge if an OCP/ILIMIT fault had latched it off. */
    fun clearFault() {
        writeReg(REG_FAULT, FAULT_CLEAR)
    }

    // -------------------------------------------------------------------------
    // Interrupt API (Level 1 — FAULTn)
    // -------------------------------------------------------------------------

    /**
     * Subscribe to fault interrupts. The fault is not cleared.
     *
     * @param callback called with the [readFault] result on each new fault
     * @param intPin FAULTn pin to arm (falling edge), or `null` to force the 5 ms polling fallback
     */
    fun onInterrupt(callback: (Fault) -> Unit, intPin: InputPin? = connection.intPin()) {
        offInterrupt()
        this.callback = callback
        if (intPin != null) {
            this.intPin = intPin
            intPin.onEdge(edgeHandler, EdgeTrigger.FALLING)
        } else {
            startPolling()
        }
    }

    /** Unsubscribe and stop delivery. */
    fun offInterrupt() {
        callback = null
        intPin?.offEdge(edgeHandler)
        intPin = null
        stopPolling()
    }

    /** Read the fault status — equivalent to [readFault], does not clear. */
    fun pollInterrupt(): Fault = readFault()

    private fun startPolling() {
        polling = true
        pollThread = Thread({
            var wasFault = false
            while (polling) {
                try {
                    val status = pollInterrupt()
                    if (status.fault && !wasFault) callback?.invoke(status)
                    wasFault = status.fault
                } catch (_: IOException) {
                    // bus error; retry on the next tick
                }
                try {
                    Thread.sleep(5)
                } catch (_: InterruptedException) {
                    Thread.currentThread().interrupt()
                    return@Thread
                }
            }
        }, "drv8830-poll").apply {
            isDaemon = true
            start()
        }
    }

    private fun stopPolling() {
        polling = false
        pollThread?.interrupt()
        pollThread = null
    }

    private fun handleEdge() {
        try {
            val status = pollInterrupt()
            if (status.fault) callback?.invoke(status)
        } catch (_: IOException) {
            // bus error; wait for the next edge rather than propagating
        }
    }
}
